randhound/protocol/utils.go

```diff
@@ -0,0 +1,256 @@
+	"reflect"
+	"sort"
+	"time"
+
+	"gopkg.in/dedis/crypto.v0/abstract"
+	"gopkg.in/dedis/crypto.v0/hash"
+	"gopkg.in/dedis/crypto.v0/random"
+	"gopkg.in/dedis/crypto.v0/share/pvss"
+	"gopkg.in/dedis/crypto.v0/sign"
+	"gopkg.in/dedis/onet.v1"
+	"gopkg.in/dedis/onet.v1/network"
+)
+
+func (rh *RandHound) newSession(nodes int, groups int, purpose string, timestamp time.Time, seed []byte, clientKey abstract.Point) (*Session, error) {
+
+	var err error
+
+	if timestamp.IsZero() {
+		timestamp = time.Now().UTC()
+	}
+
+	timestamp = timestamp.UTC() // convert to UTC
+
+	if seed == nil {
+		seed = random.Bytes(rh.Suite().Hash().Size(), random.Stream)
+	}
+
+	// Shard servers
+	indices, err := Shard(rh.Suite(), seed, nodes, groups)
+	if err != nil {
+		return nil, err
+	}
+
+	// Setup group information
+	treeNodes := rh.List()
+	servers := make([][]*onet.TreeNode, groups)
+	serverKeys := make([][]abstract.Point, groups)
+	thresholds := make([]uint32, groups)
+	groupNum := make(map[int]int)
+	groupPos := make(map[int]int)
+	for i, group := range indices {
+		s := make([]*onet.TreeNode, len(group))
+		k := make([]abstract.Point, len(group))
+		for j, g := range group {
+			s[j] = treeNodes[g]
+			k[j] = treeNodes[g].ServerIdentity.Public
+			groupNum[g] = i
+			groupPos[g] = j
+		}
+		servers[i] = s
+		serverKeys[i] = k
+		thresholds[i] = uint32(len(s)/3 + 1)
+	}
+
+	// Compute session identifier
+	sid, err := sessionID(rh.Suite(), clientKey, serverKeys, indices, purpose, timestamp)
+	if err != nil {
+		return nil, err
+	}
+
+	// Setup session
+	session := &Session{
+		nodes:      nodes,
+		groups:     groups,
+		purpose:    purpose,
+		time:       timestamp,
+		seed:       seed,
+		clientKey:  clientKey,
+		sid:        sid,
+		servers:    servers,
+		serverKeys: serverKeys,
+		indices:    indices,
+		thresholds: thresholds,
+		groupNum:   groupNum,
+		groupPos:   groupPos,
+	}
+
+	return session, nil
+}
+
+func sessionID(suite abstract.Suite, clientKey abstract.Point, serverKeys [][]abstract.Point, indices [][]int, purpose string, timestamp time.Time) ([]byte, error) {
+	// Setup some buffers
+	keyBuf := new(bytes.Buffer)
+	idxBuf := new(bytes.Buffer)
+	miscBuf := new(bytes.Buffer)
+
+	// Process client key
+	cb, err := clientKey.MarshalBinary()
+	if err != nil {
+		return nil, err
+	}
+	if _, err := keyBuf.Write(cb); err != nil {
+		return nil, err
+	}
+
+	// Process server keys and group indices
+	for i := range serverKeys {
+		for j := range serverKeys[i] {
+			kb, err := serverKeys[i][j].MarshalBinary()
+			if err != nil {
+				return nil, err
+			}
+			if _, err := keyBuf.Write(kb); err != nil {
+				return nil, err
+			}
+			if err := binary.Write(idxBuf, binary.LittleEndian, uint32(indices[i][j])); err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	// Process purpose string
+	if _, err := miscBuf.WriteString(purpose); err != nil {
+		return nil, err
+	}
+
+	// Process time stamp
+	timestamp = timestamp.UTC()
+	t, err := timestamp.MarshalBinary()
```

Comment:
I vote for using `time.Unix()` instead of doing a `MarshalBinary()` and `UTC()` everywhere. If you use `time.Unix()`, I think you can get rid of all `.UTC()`-calls all over the place.